src/libmcrx.c

```diff
@@ -1055,3 +1075,429 @@ mcrx_is_system_error(enum mcrx_error_code err) {
+    }
+    struct mcrx_subscription *sub;
+    LIST_FOREACH(sub, &ctx->subs_head, sub_entries)
+    {
+      if (sub->joined) {
+        // find new entry based on the new mnat context
+        struct mcrx_mnat_entry *new_entry =
+            mcrx_mnatmap_find_or_alloc_entry_from_subscription(sub,
+                ctx->mnat_map);
+        if (new_entry == NULL
+            || !mcrx_mnatmap_entry_local_equal(sub->mnat_entry, new_entry)) {
+          mcrx_subscription_leave(sub);
+          sub->mnat_entry = new_entry;
+          if (sub->mnat_entry
+              && !mcrx_mnatmap_entry_unresolved(sub->mnat_entry)) {
+            // rejoin with new entry
+            info(sub->ctx, "subscription %p rejoin new mnat entry %p\n", (void *)sub, (void*)sub->mnat_entry);
+            enum mcrx_error_code err = mcrx_subscription_native_join(sub);
+            if (err == MCRX_ERR_OK) {
+              ctx->live_subs++;
+            }
+          }
+        }
+      }
+    }
+  } else {
+    struct mcrx_mnatmap *clone_context;
+    enum mcrx_error_code ret = mcrx_mnatmap_clone(mnatmap, &clone_context);
+    if (ret != MCRX_ERR_OK) {
+      return ret;
+    }
+    struct mcrx_subscription *sub;
+    LIST_FOREACH(sub, &ctx->subs_head, sub_entries) {
+      // find new entry based on the new mnat context
+      struct mcrx_mnat_entry *new_entry = mcrx_mnatmap_find_or_alloc_entry_from_subscription(sub, clone_context);
+      if (sub->joined) {
+        if (new_entry == NULL || !mcrx_mnatmap_entry_local_equal(sub->mnat_entry, new_entry)) {
+          mcrx_subscription_leave(sub);
+          sub->mnat_entry = new_entry;
+          if (sub->mnat_entry
+              && !mcrx_mnatmap_entry_unresolved(sub->mnat_entry)) {
+            // rejoin with new entry
+            info(sub->ctx, "subscription %p rejoin new mnat entry %p\n", (void *)sub, (void*)sub->mnat_entry);
+            enum mcrx_error_code err = mcrx_subscription_native_join(sub);
+            if (err == MCRX_ERR_OK) {
+              ctx->live_subs++;
+            }
+          }
+        }
+      } else {
+        sub->mnat_entry = new_entry;
+        if (sub->mnat_entry && !mcrx_mnatmap_entry_unresolved(sub->mnat_entry)) {
+          enum mcrx_error_code err = mcrx_subscription_native_join(sub);
+          if (err == MCRX_ERR_OK) {
+            ctx->live_subs++;
+          }
+        }
+      }
+    }
+    // replace the run time mnat context
+    mcrx_mnatmap_unref(ctx->mnat_map);
+    ctx->mnat_map = clone_context;
+  }
+
+  info(ctx, "mnat context %p apply\n", (void* )mnatmap);
+
+  return MCRX_ERR_OK;
+}
+
+/**
+ * mcrx_mnatmap_get_mapping:
+ *
+ * Retrieve the mcrx mnat local addresses based on global source and group addresses.
+ *
+ **/
+MCRX_EXPORT enum mcrx_error_code mcrx_mnatmap_get_mapping(
+    struct mcrx_mnatmap *mnatmap, const struct mcrx_source_group_addrs *global_address,
+    struct mcrx_source_group_addrs *local_address) {
+  if (global_address == NULL || local_address == NULL) {
```

Comment:
considering how we're changing unresolved to mean the local addr_type is UNKNOWN, should this function change?  Maybe we need an error code to indicate unresolved, or should NOT_FOUND be the same as unresolved?